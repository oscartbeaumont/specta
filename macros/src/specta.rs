// inspired by https://github.com/tauri-apps/tauri/blob/2901145c497299f033ba7120af5f2e7ead16c75a/core/tauri-macros/src/command/handler.rs

use quote::quote;
use syn::{parse_macro_input, FnArg, ItemFn, Visibility};

use crate::utils::format_fn_wrapper;

pub fn attribute(item: proc_macro::TokenStream) -> syn::Result<proc_macro::TokenStream> {
    let function = parse_macro_input::parse::<ItemFn>(item)?;
    let wrapper = format_fn_wrapper(&function.sig.ident);

    let visibility = &function.vis;
    let maybe_macro_export = match &visibility {
        Visibility::Public(_) => quote!(#[macro_export]),
        _ => Default::default(),
    };

    let function_name = &function.sig.ident;

    let arg_names = function.sig.inputs.iter().map(|input| match input {
        FnArg::Receiver(_) => unreachable!("Commands cannot take 'self'"),
        FnArg::Typed(arg) => &arg.pat,
    });

    let arg_signatures = function.sig.inputs.iter().map(|_| quote!(_));

    Ok(quote! {
        #function

        #maybe_macro_export
        #[doc(hidden)]
        macro_rules! #wrapper {
            (@name) => { stringify!(#function_name) };
            (@arg_names) => { &[#(stringify!(#arg_names)),* ] };
            (@signature) => { fn(#(#arg_signatures),*) -> _ };
        }

        // allow the macro to be resolved with the same path as the function
        #[allow(unused_imports)]
        #visibility use #wrapper;
    }
    .into())
}
