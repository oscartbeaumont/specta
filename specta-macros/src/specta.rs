// inspired by https://github.com/tauri-apps/tauri/blob/2901145c497299f033ba7120af5f2e7ead16c75a/core/tauri-macros/src/command/handler.rs

use quote::{quote, ToTokens};
use syn::{parse_macro_input, FnArg, ItemFn, Pat, Visibility};

use crate::utils::{format_fn_wrapper, parse_attrs};

pub fn attribute(item: proc_macro::TokenStream) -> syn::Result<proc_macro::TokenStream> {
    let crate_ref = quote!(specta);
    let function = parse_macro_input::parse::<ItemFn>(item)?;
    let wrapper = format_fn_wrapper(&function.sig.ident);

    // While using wasm_bindgen and Specta is rare, this should make the DX nicer.
    if function.sig.unsafety.is_some()
        && function
            .sig
            .ident
            .to_string()
            .starts_with("__wasm_bindgen_generated")
    {
        return Err(syn::Error::new_spanned(
            function.sig.ident,
            "specta: You must apply the #[specta] macro before the #[wasm_bindgen] macro",
        ));
    }

    let visibility = &function.vis;
    let maybe_macro_export = match &visibility {
        Visibility::Public(_) => {
            quote!(#[macro_export])
        }
        _ => Default::default(),
    };

    let function_name = &function.sig.ident;
    let function_name_str = function_name.to_string();
    let function_asyncness = match function.sig.asyncness {
        Some(_) => true,
        None => false,
    };

    let arg_names = function.sig.inputs.iter().map(|input| match input {
        FnArg::Receiver(_) => unreachable!("Commands cannot take 'self'"),
        FnArg::Typed(arg) => match &*arg.pat {
            Pat::Ident(ident) => ident.ident.to_token_stream(),
            Pat::Macro(m) => m.mac.tokens.to_token_stream(),
            Pat::Struct(s) => s.path.to_token_stream(),
            Pat::Slice(s) => s.attrs[0].to_token_stream(),
            Pat::Tuple(s) => s.elems[0].to_token_stream(),
            _ => unreachable!("Commands must take named arguments"),
        },
    });

    let arg_signatures = function.sig.inputs.iter().map(|_| quote!(_));

    let mut attrs = parse_attrs(&function.attrs)?;
    let common = crate::r#type::attr::CommonAttr::from_attrs(&mut attrs)?;

    let deprecated = common.deprecated_as_tokens(&crate_ref);
    let docs = common.doc;

    let no_return_type = match function.sig.output {
        syn::ReturnType::Default => true,
        syn::ReturnType::Type(_, _) => false,
    };

    Ok(quote! {
        #function

        #maybe_macro_export
        #[doc(hidden)] // We take in `$function` from the invocation so we have `name::<concrete_generics_types>`
        macro_rules! #wrapper {
            (@export_fn; $function:path) => {
                fn export(type_map: &mut #crate_ref::TypeMap) -> #crate_ref::function::FunctionDataType {
                    specta::internal::get_fn_datatype(
                        $function as fn(#(#arg_signatures),*) -> _,
                        #function_asyncness,
                        #function_name_str.into(),
                        type_map,
                        &[#(stringify!(#arg_names).into()),* ],
                        std::borrow::Cow::Borrowed(#docs),
                        #deprecated,
                        #no_return_type,
                    )
                }
            }
        }

        // allow the macro to be resolved with the same path as the function
        #[allow(unused_imports)]
        #visibility use #wrapper;
    }
    .into())
}
