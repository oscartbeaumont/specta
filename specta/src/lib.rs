#![doc = include_str!("./docs.md")]
#![forbid(unsafe_code)]
#![warn(clippy::all, clippy::unwrap_used, clippy::panic)] // TODO: missing_docs
#![allow(clippy::module_inception)]
#![cfg_attr(docsrs2, feature(doc_cfg))]
#![doc(
    html_logo_url = "https://github.com/oscartbeaumont/specta/raw/main/.github/logo-128.png",
    html_favicon_url = "https://github.com/oscartbeaumont/specta/raw/main/.github/logo-128.png"
)]

#[doc(hidden)]
pub mod internal;

/// Types related to working with [`DataType`](crate::DataType). Exposed for advanced users.
pub mod datatype;
/// Support for exporting Rust functions.
#[cfg(feature = "function")]
#[cfg_attr(docsrs2, doc(cfg(feature = "function")))]
pub mod function;
/// Contains [`Type`] and everything related to it, including implementations and helper macros
pub mod r#type;

#[doc(hidden)] // TODO: Should we actually do this? I think not
pub use datatype::*;
pub use r#type::*;

/// Implements [`Type`] for a given struct or enum.
///
/// ## Example
///
/// ```rust
/// use specta::Type;
///
/// // Use it on structs
/// #[derive(Type)]
/// pub struct MyCustomStruct {
///     pub name: String,
/// }
///
/// #[derive(Type)]
/// pub struct MyCustomStruct2(String, i32, bool);
///
/// // Use it on enums
/// #[derive(Type)]
/// pub enum MyCustomType {
///     VariantOne,
///     VariantTwo(String, i32),
///     VariantThree { name: String, age: i32 },
/// }
/// ```
pub use specta_macros::Type;

/// Prepares a function to have its types extracted using [`fn_datatype`]
///
/// ## Example
///
/// ```rust
/// #[specta::specta]
/// fn my_function(arg1: i32, arg2: bool) -> &'static str {
///     "Hello World"
/// }
/// ```
#[cfg(feature = "function")]
#[cfg_attr(docsrs2, doc(cfg(feature = "function")))]
pub use specta_macros::specta;

#[cfg(doctest)]
doc_comment::doctest!("../../README.md");
