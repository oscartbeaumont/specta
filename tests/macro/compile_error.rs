//! This file is run with the `trybuild` crate to assert compilation errors in the Specta macros.

use specta::Type;

// Invalid inflection
#[derive(Type)]
#[serde(rename_all = "camelCase123")]
pub enum Demo2 {}

// Specta doesn't support Trait objects
#[derive(Type)]
pub struct Error {
    pub(crate) cause: Option<Box<dyn std::error::Error + Send + Sync>>,
}

// Enums can only flatten if
// at least one of their variants can flatten

#[derive(Type)]
enum UnitExternal {
    Unit,
}

#[derive(Type)]
enum UnnamedMultiExternal {
    UnnamedMulti(String, String),
}

#[derive(Type)]
struct FlattenExternal {
    #[serde(flatten)]
    unit: UnitExternal,
    #[serde(flatten)]
    unnamed_multi: UnnamedMultiExternal,
}

#[derive(Type)]
#[serde(untagged)]
enum UnnamedUntagged {
    Unnamed(String),
}

#[derive(Type)]
#[serde(untagged)]
enum UnnamedMultiUntagged {
    Unnamed(String, String),
}

#[derive(Type)]
struct FlattenUntagged {
    #[serde(flatten)]
    unnamed: UnnamedUntagged,
    #[serde(flatten)]
    unnamed_multi: UnnamedMultiUntagged,
}

// Adjacent can always flatten

#[derive(Type)]
#[serde(tag = "tag")]
enum UnnamedInternal {
    Unnamed(String),
}

// Internal can't be used with unnamed multis

#[derive(Type)]
struct FlattenInternal {
    #[serde(flatten)]
    unnamed: UnnamedInternal,
}

// Invalid attributes
#[derive(Type)]
#[specta(noshot = true)]
struct InvalidAttrs1;

#[derive(Type)]
#[specta(noshot)]
struct InvalidAttrs2;

#[derive(Type)]
struct InvalidAttrs3 {
    #[specta(noshot = true)]
    a: String,
}

#[derive(Type)]
struct InvalidAttrs4 {
    #[specta(noshot)]
    a: String,
}

#[derive(Type)]
#[specta(transparent)]
pub enum TransparentEnum {}

// TODO: https://docs.rs/trybuild/latest/trybuild/#what-to-test
