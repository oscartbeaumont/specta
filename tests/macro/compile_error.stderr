error: specta: found string literal containing an unsupported inflection
 --> tests/macro/compile_error.rs:7:22
  |
7 | #[serde(rename_all = "camelCase123")]
  |                      ^^^^^^^^^^^^^^

error: specta: trait objects are not currently supported.
  --> tests/macro/compile_error.rs:13:34
   |
13 |     pub(crate) cause: Option<Box<dyn std::error::Error + Send + Sync>>,
   |                                  ^^^

error: specta: Found unsupported container attribute 'noshot'
  --> tests/macro/compile_error.rs:75:10
   |
75 | #[specta(noshot = true)]
   |          ^^^^^^

error: specta: Found unsupported container attribute 'noshot'
  --> tests/macro/compile_error.rs:79:10
   |
79 | #[specta(noshot)]
   |          ^^^^^^

error: specta: Found unsupported field attribute 'noshot'
  --> tests/macro/compile_error.rs:84:14
   |
84 |     #[specta(noshot = true)]
   |              ^^^^^^

error: specta: Found unsupported field attribute 'noshot'
  --> tests/macro/compile_error.rs:90:14
   |
90 |     #[specta(noshot)]
   |              ^^^^^^

error: #[specta(transparent)] is not allowed on an enum
  --> tests/macro/compile_error.rs:96:5
   |
96 | pub enum TransparentEnum {}
   |     ^^^^

error[E0601]: `main` function not found in crate `$CRATE`
  --> tests/macro/compile_error.rs:96:28
   |
96 | pub enum TransparentEnum {}
   |                            ^ consider adding a `main` function to `$DIR/tests/macro/compile_error.rs`

error[E0277]: the trait bound `UnitExternal: specta::Flatten` is not satisfied
  --> tests/macro/compile_error.rs:32:11
   |
32 |     unit: UnitExternal,
   |           ^^^^^^^^^^^^ the trait `specta::Flatten` is not implemented for `UnitExternal`
   |
   = help: the following other types implement trait `specta::Flatten`:
             Arc<T>
             BTreeMap<K, V>
             Box<T>
             Cell<T>
             Duration
             FlattenExternal
             FlattenInternal
             FlattenUntagged
           and $N others
note: required by a bound in `_::<impl specta::Type for FlattenExternal>::inline::validate_flatten`
  --> tests/macro/compile_error.rs:29:10
   |
29 | #[derive(Type)]
   |          ^^^^ required by this bound in `validate_flatten`
   = note: this error originates in the derive macro `Type` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0277]: the trait bound `UnnamedMultiExternal: specta::Flatten` is not satisfied
  --> tests/macro/compile_error.rs:34:20
   |
34 |     unnamed_multi: UnnamedMultiExternal,
   |                    ^^^^^^^^^^^^^^^^^^^^ the trait `specta::Flatten` is not implemented for `UnnamedMultiExternal`
   |
   = help: the following other types implement trait `specta::Flatten`:
             Arc<T>
             BTreeMap<K, V>
             Box<T>
             Cell<T>
             Duration
             FlattenExternal
             FlattenInternal
             FlattenUntagged
           and $N others
note: required by a bound in `_::<impl specta::Type for FlattenExternal>::inline::validate_flatten`
  --> tests/macro/compile_error.rs:29:10
   |
29 | #[derive(Type)]
   |          ^^^^ required by this bound in `validate_flatten`
   = note: this error originates in the derive macro `Type` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0277]: the trait bound `UnnamedUntagged: specta::Flatten` is not satisfied
  --> tests/macro/compile_error.rs:52:14
   |
52 |     unnamed: UnnamedUntagged,
   |              ^^^^^^^^^^^^^^^ the trait `specta::Flatten` is not implemented for `UnnamedUntagged`
   |
   = help: the following other types implement trait `specta::Flatten`:
             Arc<T>
             BTreeMap<K, V>
             Box<T>
             Cell<T>
             Duration
             FlattenExternal
             FlattenInternal
             FlattenUntagged
           and $N others
note: required by a bound in `_::<impl specta::Type for FlattenUntagged>::inline::validate_flatten`
  --> tests/macro/compile_error.rs:49:10
   |
49 | #[derive(Type)]
   |          ^^^^ required by this bound in `validate_flatten`
   = note: this error originates in the derive macro `Type` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0277]: the trait bound `UnnamedMultiUntagged: specta::Flatten` is not satisfied
  --> tests/macro/compile_error.rs:54:20
   |
54 |     unnamed_multi: UnnamedMultiUntagged,
   |                    ^^^^^^^^^^^^^^^^^^^^ the trait `specta::Flatten` is not implemented for `UnnamedMultiUntagged`
   |
   = help: the following other types implement trait `specta::Flatten`:
             Arc<T>
             BTreeMap<K, V>
             Box<T>
             Cell<T>
             Duration
             FlattenExternal
             FlattenInternal
             FlattenUntagged
           and $N others
note: required by a bound in `_::<impl specta::Type for FlattenUntagged>::inline::validate_flatten`
  --> tests/macro/compile_error.rs:49:10
   |
49 | #[derive(Type)]
   |          ^^^^ required by this bound in `validate_flatten`
   = note: this error originates in the derive macro `Type` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0277]: the trait bound `UnnamedInternal: specta::Flatten` is not satisfied
  --> tests/macro/compile_error.rs:70:14
   |
70 |     unnamed: UnnamedInternal,
   |              ^^^^^^^^^^^^^^^ the trait `specta::Flatten` is not implemented for `UnnamedInternal`
   |
   = help: the following other types implement trait `specta::Flatten`:
             Arc<T>
             BTreeMap<K, V>
             Box<T>
             Cell<T>
             Duration
             FlattenExternal
             FlattenInternal
             FlattenUntagged
           and $N others
note: required by a bound in `_::<impl specta::Type for FlattenInternal>::inline::validate_flatten`
  --> tests/macro/compile_error.rs:67:10
   |
67 | #[derive(Type)]
   |          ^^^^ required by this bound in `validate_flatten`
   = note: this error originates in the derive macro `Type` (in Nightly builds, run with -Z macro-backtrace for more info)
